# E-commerce Return Risk Prediction & Return Reduction Analytics

## Phase 2: Data Preparation

### Objective
The objective of this notebook is to clean and prepare the raw e-commerce order and return dataset for downstream analytics, machine learning, and dashboard development.

### Key Tasks
- Load and inspect the dataset
- Assess data quality
- Validate missing values
- Convert date fields
- Create analytical features
- Prepare the target variable
- Export a cleaned dataset for analysis

### Dataset
E-commerce Order & Return Dataset

### Tools Used
- Python
- Pandas
- NumPy

## 1. Import Required Libraries

The first step is to import the libraries required for data manipulation and analysis.

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

print("Libraries imported successfully.")

Libraries imported successfully.


## 2. Load Dataset

In [3]:
# Load dataset

df = pd.read_csv(r"C:\Users\krish\Downloads\ecommerce_returns.csv")

# Dataset dimensions
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

# Preview first 5 rows
df.head()

Rows: 34500
Columns: 19


,order_id,customer_id,product_id,category,price,discount,quantity,payment_method,order_date,delivered_date,region,returned,request_date,return_reason,total_amount,shipping_cost,profit_margin,customer_age,customer_gender
0,O100000,C17270,P234890,Home,164.08,0.15,1,Credit Card,23/12/2023,27/12/2023,West,No,NaN,NaN,139.47,7.88,31.17,60,Female
1,O100001,C17603,P228204,Grocery,24.73,0.00,1,Credit Card,3/4/2025,9/4/2025,South,No,NaN,NaN,24.73,4.60,-2.62,37,Male
2,O100002,C10860,P213892,Electronics,175.58,0.05,1,Credit Card,8/10/2024,12/10/2024,North,No,NaN,NaN,166.80,6.58,13.44,34,Male
3,O100003,C15390,P208689,Electronics,63.67,0.00,1,UPI,14/9/2024,20/9/2024,South,No,NaN,NaN,63.67,5.50,2.14,21,Female
4,O100004,C15226,P228063,Home,16.33,0.15,1,COD,21/12/2024,27/12/2024,East,No,NaN,NaN,13.88,2.74,1.15,39,Male


## 3. Dataset Structure Inspection

In [4]:
# Dataset structure

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 34500 entries, 0 to 34499
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_id         34500 non-null  str    
 1   customer_id      34500 non-null  str    
 2   product_id       34500 non-null  str    
 3   category         34500 non-null  str    
 4   price            34500 non-null  float64
 5   discount         34500 non-null  float64
 6   quantity         34500 non-null  int64  
 7   payment_method   34500 non-null  str    
 8   order_date       34500 non-null  str    
 9   delivered_date   34500 non-null  str    
 10  region           34500 non-null  str    
 11  returned         34500 non-null  str    
 12  request_date     1903 non-null   str    
 13  return_reason    1903 non-null   str    
 14  total_amount     34500 non-null  float64
 15  shipping_cost    34500 non-null  float64
 16  profit_margin    34500 non-null  float64
 17  customer_age     34500 

### Key Observations

The dataset contains **34,500 records** and **19 columns**, representing order, customer, product, financial, and return-related information.

#### Data Type Assessment

- 12 categorical (`object`) columns
- 5 numerical (`float`) columns
- 2 numerical (`integer`) columns

#### Date Columns

The following columns are currently stored as strings and will require conversion to datetime format:

- `order_date`
- `delivered_date`
- `request_date`

Converting these fields will enable time-based analysis and feature engineering such as delivery duration.

#### Missing Values

Two columns contain missing values:

- `request_date`
- `return_reason`

Both columns have 1,903 non-null records, which matches the number of returned orders in the dataset. Therefore, these missing values are expected and represent orders that were not returned.

These values will be retained as they reflect the underlying business process rather than a data quality issue.

#### Initial Assessment

The dataset appears structurally sound and contains all key information required for return analytics, financial impact analysis, and predictive modeling.

## 4. Missing Value Analysis

In [5]:
# Missing value analysis

missing_values = pd.DataFrame({
    'Missing Values': df.isnull().sum(),
    'Missing Percentage': round((df.isnull().sum()/len(df))*100, 2)
})

missing_values[missing_values['Missing Values'] > 0] \
    .sort_values(by='Missing Values', ascending=False)

,Missing Values,Missing Percentage
request_date,32597,94.48
return_reason,32597,94.48


### Missing Value Findings

The missing value analysis shows that only two columns contain null values:

- `request_date`
- `return_reason`

Both columns have approximately **94.48% missing values**, corresponding to **32,597 orders**.

This is expected because these fields are only populated when a customer initiates a return request. Since only **1,903 orders were returned**, non-returned orders naturally do not have a return request date or return reason.

Therefore, these missing values represent valid business behavior rather than data quality issues and will be retained without imputation.

### Decision

- No missing value treatment is required.
- `request_date` and `return_reason` will be preserved for return-specific analysis.

## 5. Duplicate Record Analysis

In [7]:
# Check duplicate rows

duplicate_rows = df.duplicated().sum()

# Check duplicate order IDs

duplicate_order_ids = df['order_id'].duplicated().sum()

print(f"Duplicate Rows: {duplicate_rows}")
print(f"Duplicate Order IDs: {duplicate_order_ids}")

Duplicate Rows: 0
Duplicate Order IDs: 0


No duplicate records were identified within the dataset. This indicates that each order is uniquely represented and no data removal is required at this stage.


## 6. Date Conversion

In [9]:
# Convert date columns using the correct format

date_columns = [
    'order_date',
    'delivered_date',
    'request_date'
]

for col in date_columns:
    df[col] = pd.to_datetime(
        df[col],
        format='%d/%m/%Y',
        errors='coerce'
    )

df[date_columns].dtypes

order_date        datetime64[us]
delivered_date    datetime64[us]
request_date      datetime64[us]
dtype: object

In [10]:
df[['order_date', 'delivered_date', 'request_date']].head()

,order_date,delivered_date,request_date
0,2023-12-23,2023-12-27,NaT
1,2025-04-03,2025-04-09,NaT
2,2024-10-08,2024-10-12,NaT
3,2024-09-14,2024-09-20,NaT
4,2024-12-21,2024-12-27,NaT


## 7. Feature Engineering: Delivery Duration

To evaluate delivery performance, a new feature called `delivery_days` is created.

The feature represents the number of days between order placement and product delivery.

In [11]:
# Create delivery duration feature

df['delivery_days'] = (
    df['delivered_date'] - df['order_date']
).dt.days

# Summary statistics

df['delivery_days'].describe()

count    34500.000000
mean         4.814203
std          1.242141
min          3.000000
25%          4.000000
50%          5.000000
75%          6.000000
max         13.000000
Name: delivery_days, dtype: float64

#### Interpretation

Delivery durations appear consistent across the dataset with no negative or unrealistic values observed.

Although most deliveries occur within a narrow range, longer delivery durations may influence customer satisfaction and contribute to higher return likelihood.

## 8. Feature Engineering: Customer Age Groups
Customer age is currently available as a numerical variable.

To improve interpretability and support customer segmentation analysis, age groups will be created.

In [13]:
# Age distribution summary

df['customer_age'].describe()

count    34500.000000
mean        43.474377
std         14.980682
min         18.000000
25%         31.000000
50%         43.000000
75%         56.000000
max         69.000000
Name: customer_age, dtype: float64

In [14]:
# Create age groups

bins = [18, 25, 35, 45, 60, 100]

labels = [
    '18-25',
    '26-35',
    '36-45',
    '46-60',
    '60+'
]

df['age_group'] = pd.cut(
    df['customer_age'],
    bins=bins,
    labels=labels,
    include_lowest=True
)

# Verify distribution

df['age_group'].value_counts().sort_index()

age_group
18-25    5298
26-35    6678
36-45    6678
46-60    9901
60+      5945
Name: count, dtype: int64

#### Interpretation
The customer base is distributed across all age groups, with the 46–60 segment representing the largest share of customers.

The resulting age groups provide a more interpretable representation of customer demographics and will support return behavior analysis across different customer segments.

## 9. Target Variable Preparation

The `returned` column currently contains categorical values:

- Yes
- No

For machine learning purposes, the target variable will be encoded as a binary field:

| Original Value | Encoded Value |
|---------------|--------------|
| Yes | 1 |
| No | 0 |

This transformation enables the use of classification algorithms while preserving the original business meaning.

In [16]:
# Examine target distribution before encoding

df['returned'].value_counts()

returned
No     32597
Yes     1903
Name: count, dtype: int64

#### Return Rate

Overall Return Rate ≈ 5.52%

#### Interpretation

Only a small percentage of orders are returned, indicating a significant class imbalance.

This is an important consideration for predictive modeling because a model that predicts every order as "Not Returned" would achieve high accuracy while providing little business value.

In [17]:
# Encode target variable

df['returned'] = df['returned'].map({
    'No': 0,
    'Yes': 1
})

# Verify conversion

df['returned'].value_counts()

returned
0    32597
1     1903
Name: count, dtype: int64

## 10. Final Dataset Validation

In [18]:
# Dataset overview

print("Shape:", df.shape)

print("\nColumns:")
print(df.columns.tolist())

print("\nData Types:")
print(df.dtypes)

Shape: (34500, 21)

Columns:
['order_id', 'customer_id', 'product_id', 'category', 'price', 'discount', 'quantity', 'payment_method', 'order_date', 'delivered_date', 'region', 'returned', 'request_date', 'return_reason', 'total_amount', 'shipping_cost', 'profit_margin', 'customer_age', 'customer_gender', 'delivery_days', 'age_group']

Data Types:
order_id                      str
customer_id                   str
product_id                    str
category                      str
price                     float64
discount                  float64
quantity                    int64
payment_method                str
order_date         datetime64[us]
delivered_date     datetime64[us]
region                        str
returned                    int64
request_date       datetime64[us]
return_reason                 str
total_amount              float64
shipping_cost             float64
profit_margin             float64
customer_age                int64
customer_gender               str
deliv

In [19]:
# Missing value check

df.isnull().sum()

order_id               0
customer_id            0
product_id             0
category               0
price                  0
discount               0
quantity               0
payment_method         0
order_date             0
delivered_date         0
region                 0
returned               0
request_date       32597
return_reason      32597
total_amount           0
shipping_cost          0
profit_margin          0
customer_age           0
customer_gender        0
delivery_days          0
age_group              0
dtype: int64

# Phase 2 Conclusion

The raw e-commerce dataset was successfully transformed into an analysis-ready dataset.

### Completed Activities

- Loaded and inspected the dataset
- Assessed data quality and missing values
- Verified absence of duplicate records
- Converted date fields to datetime format
- Engineered delivery duration feature (`delivery_days`)
- Created customer segmentation feature (`age_group`)
- Encoded the target variable (`returned`)
- Validated dataset integrity after transformations

### Final Dataset

- Records: 34,500
- Features: 21
- Target Variable: `returned`

### Outcome

The dataset is now ready for exploratory analysis, SQL-based business analysis, machine learning model development, and Power BI dashboard creation.

In [21]:
# Export cleaned dataset

df.to_csv(
    "ecommerce_returns_clean.csv",
    index=False
)

print("Clean dataset exported successfully.")

Clean dataset exported successfully.
